bin/weewx/gettext.py

```diff
@@ -0,0 +1,123 @@
+  parameter page. That is especially useful in include files that
+  are included in multiple templates. 
+
+  Examples:
+  
+  Assume the localization file to be:
+  
+  [Texts]
+    Key1 = Text1
+    [[index]]
+      Title = Text2
+    [[othertemplate]]
+      Title = Text3
+  
+  With that file is:
+  
+  $gettext("Key1")           ==> Text1
+  
+  $gettext("Title",$page)    ==> is Text2 if included in template "index"
+                                 and Text3 if included in template 
+                                 "othertemplate"
+                                
+  $gettext(page=$page).Title ==> same as before
+  
+  You can use "speaking" key names. If they contain whitespace they need
+  to be enclosed in quotes in the localization file
+  
+  If the value contains commas you must enclose it in single or double
+  quotes. Otherwise an array instead of a string is returned.
+  
+"""
+
+from weewx.cheetahgenerator import SearchList
+from weeutil.weeutil import KeyDict
+import weeutil.config
+
+# Test for new-style weewx v4 logging by trying to import weeutil.logger
+import weeutil.logger
+import logging
+
+log = logging.getLogger(__name__)
+
+
+class Gettext(SearchList):
+
+    def get_extension_list(self,timespan,db_lookup):
+            
+        def locale_label(key='',page=''):
+            """ $gettext()
+            
+                key: key to look up for
+                page: section name  
+                
+            """
+            _text_dict = self.generator.skin_dict.get('Texts',{})
+
+            # get the appropriate section for page if defined
+            # Note: no hierarchy here
+            if page:
+                if page not in _text_dict:
+                    log.error("could not find section [Texts][[%s]] for report %s" % (page,self.generator.skin_dict.get('REPORT_NAME','unknown')))
+                # Note: no 'else:' here, because we need the empty dict
+                # page is specified --> get subsection "page"
+                _text_dict = _text_dict.get(page,{})
+            
+            # if key is not empty, get the value for key
+            if key:
+                # return the value for key
+                # if key not in _text_dict return key instead
+                return KeyDict(_text_dict)[key]
+                
+
+            # if key is empty but page is not return further class instance
+            if page:
+                _page_dict = weeutil.config.config_from_str('')
+                merge_dict = self.generator.skin_dict.get('Labels',{}).get('Generic',{})
+                if merge_dict:
+                    weeutil.config.merge_config(_page_dict,merge_dict)
+                merge_dict = Gettext._get_cheetah_dict(self.generator.skin_dict.get('CheetahGenerator',{}),page)
+                if merge_dict:
+                    weeutil.config.merge_config(_page_dict,merge_dict)
+                weeutil.config.merge_config(_page_dict,_text_dict)
+                return _page_dict
```

Comment:
Trying to understand something here. You allow a page to be specified with an empty key (e.g., `$gettext('', 'some_page')`). Yet, I don't see any place where this is used in the skin.

What's the purpose of this functionality?

Reply:
1. Something like `$gettext(page=$page).Title` to set a page-dependent text in include files. Especially titles, meta tags in HTML and the like 
2. Can be used when creating sitemaps. `page` can be any page, not only the actual page.
    * `$gettext(page=$page).template` - the name of the template file
    * `$gettext(page=$page).stale_age` - live time of the page
    * `$gettext(page=$page).encoding` - encoding of the page
    * other lines that are placed in this section.

And as you ask I am not sure why I merged `[Labels][[Generic]]` there. I guess it was obsoleted by some earlier change and can be removed now. I now see no sense in something like `$gettext(page=$page).outTemp` as it does not depend on pages. I would remove that merge.

Reply:
Why not go all the way, and make `gettext` act like a proper dictionary.

    $gettext[$page][template]

which in Cheetah's world can be abbreviated to

    $gettext[$page].template

This would also allow nesting:

    $gettext[$page][subpage].template
    $gettext[$page].subpage.template

Reply:
This could be implemented using a class that holds the text dictionary internally. It's like a normal dict, except that it returns the key if the key is missing. Something like:

```python
from six.moves import collections_abc


class ParallelDict(collections_abc.Mapping):

    def __init__(self, source):
        self.source = source

    def __getitem__(self, key):
        try:
            return self.source[key]
        except KeyError:
            return key

    def __len__(self):
        return self.source.__len__()

    def __iter__(self):
        for key in self.source:
            yield key
```

Reply:
I will try that.

Reply:
I changed gettext.py and replaced `$gettext()` by `$gettext[]`. It generally works. 

Actually if there is an undefined key in the second level (`$gettext["defined key"]["undefined key"]`) an error is returned that refers to `configObj` class. So I guess, I have to write a recursive function to convert the subsections to `ParallelDict` or something.

Reply:
What is the difference between `ParallelDict` and `KeyDict` regarding our needs?